File: source/modulo_core/test/cpp/test_predicate.cpp
#include <gtest/gtest.h>

#include "modulo_core/Predicate.hpp"

using namespace modulo_core;

TEST(PredicateTest, SimplePredicate) {
  auto predicate = Predicate([]() { return true; });
  EXPECT_TRUE(predicate.get_value());
  auto value = predicate.query();
  EXPECT_TRUE(value);
  EXPECT_TRUE(*value);

  EXPECT_TRUE(predicate.get_value());
  value = predicate.query();
  EXPECT_FALSE(value);

  predicate.set_predicate([]() { return false;} );
  EXPECT_FALSE(predicate.get_value());
  value = predicate.query();
  EXPECT_TRUE(value);
  EXPECT_FALSE(*value);

  EXPECT_FALSE(predicate.get_value());
  value = predicate.query();
  EXPECT_FALSE(value);
}

File: source/modulo_core/test/python/test_predicate.py
from modulo_core import Predicate


def test_simple_predicate():
    predicate = Predicate(lambda: True)
    assert predicate.get_value()
    assert predicate.query()
    assert predicate.get_value()
    assert predicate.query() is None

    predicate.set_predicate(lambda: False)
    assert predicate.get_value() is False
    assert predicate.query() is False
    assert predicate.get_value() is False
    assert predicate.query() is None


def test_predicate_change_before_query():
    predicate = Predicate(lambda: False)
    predicate.set_predicate(lambda: True)
    assert predicate.get_value()
    assert predicate.query()
    assert predicate.query() is None
